### PR TITLE
Use chart path instead of name as a map key

### PR DESCRIPTION
**Pull request by divolgin**
#### What this PR does / why we need it:

Don't use chart name as the key in an in-memory structure because charts names are not unique.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
* Fixes a bug that caused kustomize to fail if a Helm chart and one of its sub-charts had the same name
```

#### Does this PR require documentation?

NONE

### CODE REVIEW

File: pkg/pull/pull.go [hunk 1]
```diff
@@ -440,9 +440,12 @@ func Pull(upstreamURI string, pullOptions PullOptions) (string, error) {
 		chartBaseName := v.Spec.Chart.Name
 		// the helmBase may have a chart name prefix removed - we must find the base name instead of the original chart name
 		for _, helmBase := range helmBases {
-			chartName := strings.Split(helmBase.Path, "/")[len(strings.Split(helmBase.Path, "/"))-1]
-			if strings.HasSuffix(v.Spec.Chart.Name, chartName) {
+			baseDir := filepath.Dir(helmBase.Path)
+			chartName := filepath.Base(helmBase.Path)
```
Comment (sgalsaleh) [inline pkg/pull/pull.go:444]: i think the base path comes from `metadata.name` or the new `releaseName` field, not the chart name. so comparing this to `v.Spec.Chart.Name` is not accurate i believe

Comment (sgalsaleh) [reply pkg/pull/pull.go:444]: there's a `v.GetDirName()` function that should return the base directory name that was used

File: pkg/rewrite/rewrite.go [hunk 2]
```diff
@@ -237,8 +236,10 @@ func Rewrite(rewriteOptions RewriteOptions) error {
 		chartBaseName := v.Spec.Chart.Name
 		// the helmBase may have a chart name prefix removed - we must find the base name instead of the original chart name
 		for _, helmBase := range helmBases {
-			chartName := strings.Split(helmBase.Path, "/")[len(strings.Split(helmBase.Path, "/"))-1]
-			if strings.HasSuffix(chartBaseName, chartName) {
+			baseDir := filepath.Dir(helmBase.Path)
+			chartName := filepath.Base(helmBase.Path)
+			// v is a top level chart, so only top level charts directory name should match
```
Comment (sgalsaleh) [inline pkg/rewrite/rewrite.go:241]: > so only top level charts directory name should match

how is that enforced here? by `baseDir == "charts"`? isn't there a `charts` directory for each subchart as well?

Comment (divolgin) [reply pkg/rewrite/rewrite.go:241]: yes, and baseDir is the entire path. it won't match anything nested deeper.

File: pkg/pull/pull.go [hunk 3]
```diff
@@ -440,9 +440,12 @@ func Pull(upstreamURI string, pullOptions PullOptions) (string, error) {
 		chartBaseName := v.Spec.Chart.Name
 		// the helmBase may have a chart name prefix removed - we must find the base name instead of the original chart name
 		for _, helmBase := range helmBases {
-			chartName := strings.Split(helmBase.Path, "/")[len(strings.Split(helmBase.Path, "/"))-1]
-			if strings.HasSuffix(v.Spec.Chart.Name, chartName) {
+			baseDir := filepath.Dir(helmBase.Path)
```
[the remark this answers is not in the document]
Comment (divolgin) [reply pkg/pull/pull.go:443]: that leave a trailing slash in directory name, which i deemed ugly :)

File: pkg/image/builder.go [hunk 4]
```diff
@@ -102,16 +102,22 @@ func GetPrivateImages(upstreamDir string, checkedImages map[string]ImageInfo, al
...
-				return filepath.SkipDir
-			}
-
 			if info.IsDir() {
+				chartName, err := filepath.Rel(upstreamDir, path)
+				if err != nil {
+					logger.Debugf("Failed to remove prefix from %s: %v", path, err)
+					return nil
+				}
+				chartName = filepath.Join(parentHelmChartPath, chartName)
```
Comment (sgalsaleh) [inline pkg/image/builder.go:111]: shouldn't `filepath.Rel(upstreamDir, path)` already have the parent helm chart path?

File: pkg/pull/pull.go [hunk 5]
```diff
@@ -437,12 +437,15 @@ func Pull(upstreamURI string, pullOptions PullOptions) (string, error) {
...
-		chartBaseName := v.Spec.Chart.Name
+		chartBaseName := v.GetDirName()
 		// the helmBase may have a chart name prefix removed - we must find the base name instead of the original chart name
 		for _, helmBase := range helmBases {
-			chartName := strings.Split(helmBase.Path, "/")[len(strings.Split(helmBase.Path, "/"))-1]
-			if strings.HasSuffix(v.Spec.Chart.Name, chartName) {
+			baseDir := filepath.Dir(helmBase.Path)
+			chartName := filepath.Base(helmBase.Path)
+			// v is a top level chart, so only top level charts directory name should match
+			if baseDir == "charts" && strings.HasSuffix(chartBaseName, chartName) {
```
Comment (sgalsaleh) [inline pkg/pull/pull.go:446]: why not
```suggestion
			if baseDir == "charts" && chartBaseName == chartName {
```

File: pkg/rewrite/rewrite.go [hunk 6]
```diff
@@ -234,11 +233,13 @@ func Rewrite(rewriteOptions RewriteOptions) error {
...
-		chartBaseName := v.Spec.Chart.Name
+		chartBaseName := v.GetDirName()
 		// the helmBase may have a chart name prefix removed - we must find the base name instead of the original chart name
 		for _, helmBase := range helmBases {
-			chartName := strings.Split(helmBase.Path, "/")[len(strings.Split(helmBase.Path, "/"))-1]
-			if strings.HasSuffix(chartBaseName, chartName) {
+			baseDir := filepath.Dir(helmBase.Path)
+			chartName := filepath.Base(helmBase.Path)
+			// v is a top level chart, so only top level charts directory name should match
+			if baseDir == "charts" && strings.HasSuffix(chartBaseName, chartName) {
```
Comment (sgalsaleh) [inline pkg/rewrite/rewrite.go:242]: ditto

File: pkg/image/builder.go [hunk 7]
```diff
@@ -91,7 +91,7 @@ func ProcessImages(srcRegistry, destRegistry registry.RegistryOptions, appSlug s
 	return newImages, nil
 }
 
-func GetPrivateImages(upstreamDir string, checkedImages map[string]ImageInfo, allPrivate bool, dockerHubRegistry registry.RegistryOptions, useHelmInstall map[string]bool) ([]string, []k8sdoc.K8sDoc, error) {
+func GetPrivateImages(upstreamDir string, checkedImages map[string]ImageInfo, allPrivate bool, dockerHubRegistry registry.RegistryOptions, parentHelmChartPath string, useHelmInstall map[string]bool) ([]string, []k8sdoc.K8sDoc, error) {
```
Comment (sgalsaleh) [inline pkg/image/builder.go:94]: this function is not just for helm charts though, a bit confusing that there is a parameter named `parentHelmChartPath`

Comment (sgalsaleh) [reply pkg/image/builder.go:94]: maybe pass base? or convert this function to be something like
```
func (b *Base) GetPrivateImages(
```

File: pkg/image/builder.go (hunk 4 above)
Comment (divolgin) [reply pkg/image/builder.go:111]: each helm chart is treated as a separate base, so starting root will be nested inside one or more charts, and the map contains values relative to top level.

File: pkg/pull/pull.go (hunk 5 above)
Comment (divolgin) [reply pkg/pull/pull.go:446]: This makes this loop pointless.  I don't actually know what it does.  How about just removing it?

File: pkg/rewrite/rewrite.go (hunk 6 above)
Comment (divolgin) [reply pkg/rewrite/rewrite.go:242]: ditto :)

File: pkg/image/builder.go (hunk 4 above)
Comment (sgalsaleh) [reply pkg/image/builder.go:111]: ah that's why i asked if a double nested subchart would work here

File: pkg/pull/pull.go (hunk 5 above)
Comment (sgalsaleh) [reply pkg/pull/pull.go:446]: yes, i agree. we should remove it

Comment (sgalsaleh) [reply pkg/pull/pull.go:446]: it does seem pointless indeed. no objection on removing it

File: pkg/image/builder.go (hunk 7 above)
Comment (divolgin) [reply pkg/image/builder.go:94]: `maybeParentHelmChartPath`?  it's actually a bit of an invariant.  without parent chart, it's empty string, which still works with the `useHelmInstall` map

Comment (sgalsaleh) [reply pkg/image/builder.go:94]: i don't have any bright ideas. no strong objection to keep it as is. maybe aggregate all of these params in a new options struct or something

Comment (divolgin) [reply pkg/image/builder.go:94]: the name makes sense to me given existing context

**Review by sgalsaleh — APPROVED** [review]

File: pkg/tests/pull/cases/subcharts/wantResults/overlays/midstream/charts/fluent-bit/kustomization.yaml [hunk 8]
```diff
@@ -0,0 +1,25 @@
+apiVersion: kustomize.config.k8s.io/v1beta1
+bases:
+- ../../../../base/charts/fluent-bit
+commonAnnotations:
+  kots.io/app-slug: my-app
+images:
```
[the remark this answers is not in the document]
Comment (divolgin) [reply pkg/tests/pull/cases/subcharts/wantResults/overlays/midstream/charts/fluent-bit/kustomization.yaml:6]: yes, this is current behavior.  not introduced in this PR

File: pkg/tests/pull/cases/subcharts/wantResults/overlays/midstream/kustomization.yaml [hunk 9]
```diff
@@ -34,6 +34,14 @@ images:
   newName: fake-docker-proxy/busybox
 - name: library/busybox
   newName: fake-docker-proxy/busybox
+- name: cr.fluentbit.io/fluent/fluent-bit
```
[the remark this answers is not in the document]
Comment (divolgin) [reply pkg/tests/pull/cases/subcharts/wantResults/overlays/midstream/kustomization.yaml:37]: yep, same issue as the one in the parent chart.

**Pull request merged**